File: src/InfiniFrame.Js/Interop/InteropEnvelopeProtocol.cs
// ---------------------------------------------------------------------------------------------------------------------
// Imports
// ---------------------------------------------------------------------------------------------------------------------
using System.Text;
using System.Text.Json;
using InfiniFrame.Interop;

namespace InfiniFrame.Js.Interop;
// ---------------------------------------------------------------------------------------------------------------------
// Code
// ---------------------------------------------------------------------------------------------------------------------
internal static class InteropEnvelopeProtocol {
    internal const int CurrentVersion = 1;
    internal const int MaxMessageSizeBytes = 1024 * 1024;

    private static readonly JsonDocumentOptions JsonDocumentOptions = new() {
        AllowTrailingCommas = false,
        CommentHandling = JsonCommentHandling.Disallow,
        MaxDepth = 64
    };

    internal static string CreateEnvelopeMessage(string id, string? data = null) {
        ArgumentException.ThrowIfNullOrWhiteSpace(id);

        using var stream = new MemoryStream();
        using var writer = new Utf8JsonWriter(stream);

        writer.WriteStartObject();
        writer.WriteString("id", id);
        if (data is null)
            writer.WriteNull("data");
        else
            writer.WriteString("data", data);
        writer.WriteNumber("version", CurrentVersion);
        writer.WriteEndObject();
        writer.Flush();

        return Encoding.UTF8.GetString(stream.ToArray());
    }

    internal static InteropEnvelopeParseResult ParseIncomingMessage(string message) {
        if (string.IsNullOrWhiteSpace(message))
            return InteropEnvelopeParseResult.CreateFailure("Message is empty.");

        int byteCount = Encoding.UTF8.GetByteCount(message);
        if (byteCount > MaxMessageSizeBytes)
            return InteropEnvelopeParseResult.CreateFailure($"Message exceeds max size of {MaxMessageSizeBytes} bytes.");

        if (!LooksLikeJsonObject(message))
            return InteropEnvelopeParseResult.CreateFailure("Message is not a valid JSON envelope.");

        try {
            using JsonDocument jsonDocument = JsonDocument.Parse(message, JsonDocumentOptions);
            JsonElement root = jsonDocument.RootElement;
            if (root.ValueKind != JsonValueKind.Object)
                return InteropEnvelopeParseResult.CreateFailure("Envelope root must be a JSON object.");

            if (!root.TryGetProperty("id", out JsonElement idElement) || idElement.ValueKind != JsonValueKind.String)
                return InteropEnvelopeParseResult.CreateFailure("Envelope 'id' is required and must be a string.");

            string? messageId = idElement.GetString();
            if (string.IsNullOrWhiteSpace(messageId))
                return InteropEnvelopeParseResult.CreateFailure("Envelope 'id' cannot be empty.");

            if (!root.TryGetProperty("version", out JsonElement versionElement) || versionElement.ValueKind != JsonValueKind.Number)
                return InteropEnvelopeParseResult.CreateFailure("Envelope 'version' is required and must be a number.");

            if (!versionElement.TryGetInt32(out int version))
                return InteropEnvelopeParseResult.CreateFailure("Envelope 'version' must be a 32-bit integer.");

            if (version != CurrentVersion)
                return InteropEnvelopeParseResult.CreateFailure($"Unsupported envelope version '{version}'.");

            string? payload = null;
            if (root.TryGetProperty("data", out JsonElement dataElement))
                payload = ConvertDataToPayload(dataElement);

            return InteropEnvelopeParseResult.CreateSuccess(messageId, payload);
        }
        catch (JsonException) {
            return InteropEnvelopeParseResult.CreateFailure("Envelope JSON is malformed.");
        }
    }

    private static bool LooksLikeJsonObject(string message) {
        ReadOnlySpan<char> span = message.AsSpan().TrimStart();
        return !span.IsEmpty && span[0] == '{';
    }

    private static string? ConvertDataToPayload(JsonElement dataElement) {
        return dataElement.ValueKind switch {
            JsonValueKind.Null => null,
            JsonValueKind.String => dataElement.GetString(),
            _ => dataElement.GetRawText()
        };
    }
}

File: src/InfiniFrame.Js/TypeScript/Interop/InteropEnvelopeProtocol.ts
// ---------------------------------------------------------------------------------------------------------------------
// Imports
// ---------------------------------------------------------------------------------------------------------------------
import {InteropEnvelopeV1, ParsedInteropMessage, InteropParseError} from "../Contracts/IInteropEnvelope";

// ---------------------------------------------------------------------------------------------------------------------
// Code
// ---------------------------------------------------------------------------------------------------------------------
export const InteropEnvelopeVersion = 1;
export const InteropMessageMaxSizeBytes = 1024 * 1024;

export function createEnvelope(id: string, data?: unknown, channel?: string): InteropEnvelopeV1 {
    if (!id || id.trim().length === 0) {
        throw new Error("Envelope 'id' is required.");
    }

    return {
        id,
        data,
        version: InteropEnvelopeVersion,
        channel
    };
}

export function createEnvelopeMessage(id: string, data?: unknown, channel?: string): string {
    const envelope = createEnvelope(id, data, channel);

    return JSON.stringify(envelope);
}

export function parseIncomingMessage(message: string): ParsedInteropMessage | InteropParseError {
    if (!message || message.trim().length === 0) {
        return {error: "Message is empty."};
    }

    if (getUtf8ByteCount(message) > InteropMessageMaxSizeBytes) {
        return {error: `Message exceeds max size of ${InteropMessageMaxSizeBytes} bytes.`};
    }

    if (!looksLikeJsonObject(message)) {
        return parseLegacyMessage(message);
    }

    try {
        const parsed = JSON.parse(message) as unknown;
        if (!isObject(parsed)) {
            return {error: "Envelope root must be a JSON object."};
        }

        if (typeof parsed.id !== "string" || parsed.id.trim().length === 0) {
            return {error: "Envelope 'id' is required and must be a string."};
        }

        if (typeof parsed.version !== "number" || !Number.isInteger(parsed.version)) {
            return {error: "Envelope 'version' is required and must be an integer."};
        }

        if (parsed.version !== InteropEnvelopeVersion) {
            return {error: `Unsupported envelope version '${parsed.version}'.`};
        }

        const payload = convertDataToPayload(parsed.data);
        return {
            messageId: parsed.id,
            payload
        };
    } catch {
        return {error: "Envelope JSON is malformed."};
    }
}

function parseLegacyMessage(message: string): ParsedInteropMessage | InteropParseError {
    const separatorIndex = message.indexOf(";");
    const hasSeparator = separatorIndex >= 0;
    const messageId = (hasSeparator ? message.slice(0, separatorIndex) : message).trim();

    if (messageId.length === 0) {
        return {error: "Legacy message has an empty message ID."};
    }

    return {
        messageId,
        payload: hasSeparator ? message.slice(separatorIndex + 1) : undefined,
        isLegacyProtocol: true
    };
}

function convertDataToPayload(data: unknown): string | undefined {
    if (data === null || data === undefined) {
        return undefined;
    }

    if (typeof data === "string") {
        return data;
    }

    return JSON.stringify(data);
}

function looksLikeJsonObject(message: string): boolean {
    return message.replace(/^\s+/, "").startsWith("{");
}

function getUtf8ByteCount(message: string): number {
    return new TextEncoder().encode(message).length;
}

function isObject(value: unknown): value is Record<string, unknown> {
    return typeof value === "object" && value !== null;
}
